lib/BuildSystem/BuildSystem.cpp

```diff
@@ -2041,3 +2041,10 @@ bool BuildSystem::enableTracing(StringRef path,
 bool BuildSystem::build(StringRef name) {
   return static_cast<BuildSystemImpl*>(impl)->build(name);
 }
+
+void BuildSystem::cancel() {
+  if (impl) {
```

Comment:
I think we need a flag to track whether we are cancelling, to cause jobs to stop trying to enqueue work if we are trying to cancel.

We should fix the implementation to properly merge this and the `getDelegate().isCancelled()` API, so that inside the BuildSystem we only have one hook.